apps/opentelemetry/src/otel_batch_olp.erl

```diff
@@ -0,0 +1,623 @@
+               reg_name              :: atom(),
+               max_queue_size        :: max_queue_size(),
+               exporting_timeout_ms  :: otel_timeout_ms(),
+               scheduled_delay_ms    :: otel_timeout_ms(),
+               shutdown_ms           :: otel_timeout_ms(),
+               atomic_ref            :: atomics:atomic_ref(),
+               exporter_timer        :: undefined | reference(),
+               otel_signal           :: otel_signal(),
+               %% for future extensions
+               cb_module             :: module(),
+               %% Arbitrary config of the callback module
+               external_config       :: term(),
+               extra                 = [] %% Unused, for future extensions
+              }).
+
+%%--------------------------------------------------------------------
+%% Test utils
+%%--------------------------------------------------------------------
+
+-ifdef(TEST).
+-export([current_tab_to_list/1]).
+current_tab_to_list(RegName) ->
+    {_, #data{tables=Tabs, atomic_ref=AtomicRef}} = sys:get_state(RegName),
+    case ?current_tab(AtomicRef) of
+        0 -> [];
+        TabIx -> ets:tab2list(?tab_name(TabIx, Tabs))
+    end.
+-endif.
+
+%%--------------------------------------------------------------------
+%% otel_batch_olp API
+%%--------------------------------------------------------------------
+
+-spec init_conf(otel_batch_olp_config()) -> {ok, otel_batch_olp_state()} | {error, term()}.
+init_conf(#{reg_name := RegName, cb_module := _Module, otel_signal := _, exporter := _,
+            max_queue_size := _, exporting_timeout_ms := _, scheduled_delay_ms := _} = Config) ->
+    case validate_config(Config) of
+        ok ->
+            AtomicRef = atomics:new(3, [{signed, true}]),
+            {ok, Config#{reg_name => RegName,
+                          tables => {?table_1(RegName), ?table_2(RegName)},
+                          atomic_ref => AtomicRef}};
+        Err ->
+            Err
+    end.
+-spec start_link(otel_batch_olp_state(), term()) -> gen_statem:start_ret().
+start_link(#{reg_name := RegName} = OlpState, ExternalConfig) ->
+    gen_statem:start_link({local, RegName}, ?MODULE, [OlpState, ExternalConfig], []).
+
+-spec insert_signal(tuple(), otel_batch_olp_state()) -> true | dropped | {error, term()}.
+insert_signal(Record, #{atomic_ref := AtomicRef, tables := Tabs} = State) ->
+    try
+        case ?current_tab(AtomicRef) of
+            0 -> dropped;
+            CurrentTab ->
+                case ?sub_get_available(AtomicRef, CurrentTab) of
+                    Seq when Seq > 0 ->
+                        ets:insert(?tab_name(CurrentTab, Tabs), Record);
+                    0 ->
+                        %% max_queue_size is reached
+                        Res = ets:insert(?tab_name(CurrentTab, Tabs), Record),
+                        _ = force_flush(State),
+                        Res;
+                    _ ->
+                        dropped
```

Comment:
Just curious: if we have concurrent intensive writes, will this happen frequently?

Reply:
Depends on the configured max_queue_size and exporter performance/ scheduled timeout.
If write rate is higher than baych_olp/exporter can handle, then this can happen frequently. There are some test cases that simulate this.